### PR TITLE
feat(neighbors): add target rows and compiled pair functions

### PR DESCRIPTION
**Pull request by zubatyuk**
# ALCHEMI Toolkit-Ops Pull Request

## Description

Add compact-row partial neighbor-list support across JAX and PyTorch naive/cell-list bindings, including batched paths, pair-output routes, buffer validation, and fixed-shape Torch fullgraph pair-function wrappers.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Performance improvement
- [x] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Related Issues

None.

## Changes Made

- Add compact `target_indices` rows to JAX and PyTorch naive/cell-list paths, including batched variants, low-level JAX cell-list query wrappers, caller-provided compact buffer validation, and JAX batch default cell/PBC normalization.
- Add Torch `compile_pair_fn()` / `CompiledPairFn` support for fixed-shape matrix `pair_fn` routes under `torch.compile(fullgraph=True)`, with eager-only handling for raw Warp functions.
- Update neighbor-list docs and the JAX neighbor-list example for compact-row target lists, JAX buffer-shape wording, pair-function support, and cluster-tile limitations.
- Add JAX and PyTorch regression tests for compact target rows, pair outputs, JIT/fullgraph behavior, user-provided buffers, and validation errors.

## Testing

- [x] Unit tests pass locally (`make pytest`)
- [x] Linting passes (`make lint`)
- [x] New tests added for new functionality meets coverage expectations?

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [ ] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [x] I have added docstrings to new functions/classes
- [x] I have updated the documentation (if applicable)

### Support Matrix

Legend: ✅ supported, ⚠️ supported with constraints, ❌ unsupported / out of scope, N/A not applicable.

#### High-Level Neighbor-List Entry Points

| Backend | Entry points | `target_indices` compact rows | `pair_fn` eager | JIT / compile path | Notes |
| --- | --- | --- | --- | --- | --- |
| Warp | `naive`, `batch_naive` | ✅ | ✅ | N/A | Matrix and COO source rows use compact row ids for partial lists. |
| Warp | `cell_list`, `batch_cell_list` | ✅ | ✅ | N/A | Build/query split remains cacheable; partial rows are compact. |
| Warp | `cluster_tile`, `batch_cluster_tile` | ❌ | ✅ | N/A | Cluster-tile does not expose partial central rows. |
| JAX | `naive`, `batch_naive` | ✅ | ✅ | ⚠️ `jax.jit` with fixed shapes / static sizing metadata | Explicit tile strategy plus partial rows is rejected. |
| JAX | `cell_list`, `batch_cell_list` | ✅ | ✅ | ⚠️ fixed-shape atom-centric/query paths | `strategy="pair_centric" + target_indices` remains rejected; use atom-centric / auto. |
| JAX | `cluster_tile`, `batch_cluster_tile` | ❌ | ⚠️ fp32 eager-cutoff paths | ⚠️ matrix output only; COO packing is data-dependent | No partial rows. |
| PyTorch | `naive`, `batch_naive` | ✅ | ✅ raw `wp.Function` eager; ✅ `CompiledPairFn` fixed matrix | ⚠️ `torch.compile(fullgraph=True)` with fixed buffers | Raw `wp.Function` remains eager-only under fullgraph. |
| PyTorch | `cell_list`, `batch_cell_list` | ✅ | ✅ raw `wp.Function` eager; ✅ `CompiledPairFn` fixed matrix | ⚠️ `torch.compile(fullgraph=True)` with fixed buffers | Fullgraph support is matrix-only; COO packing stays eager. |
| PyTorch | `cluster_tile`, `batch_cluster_tile` | ❌ | ✅ | ❌ no `CompiledPairFn` fullgraph route | No partial rows. |

#### Low-Level / Specialized Entry Points

| Backend | Entry point | `target_indices` | Optional geometry / `pair_fn` outputs | JIT / compile notes |
| --- | --- | --- | --- | --- |
| JAX | `query_cell_list`, `batch_query_cell_list` | ✅ compact rows | ✅ `return_distances`, `return_vectors`, `pair_fn`, `pair_energies`, `pair_forces` | Uses the same fixed-shape pair-output route as the combined wrappers. |
| JAX | `build_cell_list`, `batch_build_cell_list` | ❌ build-only | ❌ build-only | Errors point users to query or combined wrappers for output features. |
| JAX | `query_cluster_tile`, `batch_query_cluster_tile` | ❌ | ⚠️ `pair_fn` on supported fp32 tile paths | No partial-row API. |
| PyTorch | `query_cell_list`, `batch_query_cell_list` | ✅ compact rows | ✅ eager raw `wp.Function`; ✅ `CompiledPairFn` fixed matrix | Fullgraph requires caller-provided fixed buffers. |
| PyTorch | cluster-tile direct wrappers | ❌ | ✅ eager | No fullgraph `CompiledPairFn` route in this PR. |

#### Torch `pair_fn` Under `torch.compile(fullgraph=True)`

| Route | Raw `wp.Function` | `compile_pair_fn(pair_fn)` / `CompiledPairFn` | Output forms | Required buffers |
| --- | --- | --- | --- | --- |
| `naive` | ❌ eager-only | ✅ no-PBC and PBC, including `target_indices` | Matrix only | Neighbor matrix/counts/shifts, pair geometry/output buffers, PBC shift metadata when needed. |
| `batch_naive` | ❌ eager-only | ✅ no-PBC and PBC, including `target_indices` | Matrix only | Batched compact/full buffers plus PBC shift metadata when needed. |
| `cell_list` | ❌ eager-only | ✅ including `target_indices` | Matrix only | Fixed neighbor, scratch/cache, geometry, and pair-output buffers. |
| `batch_cell_list` | ❌ eager-only | ✅ including `target_indices` | Matrix only | Fixed batched neighbor, scratch/cache, geometry, and pair-output buffers. |
| `cluster_tile`, `batch_cluster_tile` | ❌ eager-only under fullgraph | ❌ | N/A | Out of scope for this PR. |

> [!TIP]
> This repository uses Greptile, an AI code reviewer. Greptile will automatically review your PR and provide feedback. You can also ask Greptile questions about your code by commenting on the PR.

### CODE REVIEW

File: nvalchemiops/torch/neighbors/cell_list.py [hunk 1]
```diff
@@ -1819,13 +1975,12 @@ def cell_list(
...
         if pair_fn is not None and pair_energies is None:
             pair_energies = torch.zeros(
-                (total_atoms, max_neighbors), dtype=positions.dtype, device=device
+                (num_rows, max_neighbors), dtype=positions.dtype, device=device
             )
         if pair_fn is not None and pair_forces is None:
             pair_forces = torch.zeros(
-                (total_atoms, max_neighbors, 3), dtype=positions.dtype, device=device
+                (num_rows, max_neighbors, 3), dtype=positions.dtype, device=device
             )
```
[the remark this answers is not in the document]
Comment (zubatyuk) [reply nvalchemiops/torch/neighbors/cell_list.py:1983]: `torch.compile(fullgraph=True)` path already validates fixed buffers explicitly.

In the eager path, `CompiledPairFn` succeeds without requested distance/vector outputs for both `cell_list` and `batch_cell_list`. The optional query path accepts `None` geometry buffers.

File: nvalchemiops/torch/neighbors/_compiled_pair_fn.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+from threading import Lock
+from typing import TypeVar
+
+import warp as wp
+
+__all__ = ["CompiledPairFn", "compile_pair_fn", "is_compiled_pair_fn"]
+
+_T = TypeVar("_T")
+_COUNTER = count()
+_LOCK = Lock()
```
[the remark this answers is not in the document]
Comment (zubatyuk) [reply nvalchemiops/torch/neighbors/_compiled_pair_fn.py:32]: It is defensive. If this code is only called from single-threaded setup, the lock will never matter, but it prevents rare duplicate-registration races in multi-threaded callers/test runners.

File: examples/neighbors/05_jax_neighbor_list.py [hunk 3]
```diff
@@ -340,13 +342,54 @@ def run_compute_loop(
...
+
+partial_coo, partial_ptr, _ = cell_list(
+    positions,
+    cutoff,
+    cell,
+    pbc,
+    max_neighbors=128,
+    target_indices=target_indices,
+    return_neighbor_list=True,
+    strategy="atom_centric",
```
Comment (dallasfoster) [inline examples/neighbors/05_jax_neighbor_list.py]: Why do we call it strategy here instead of method?

File: examples/neighbors/06_pair_outputs_lj.py [hunk 4]
```diff
@@ -172,41 +187,148 @@ def lj_pair_fn(
...
+        rtol=5.0e-4,
+        atol=5.0e-6,
+    )
+    torch.testing.assert_close(
+        active_forces,
+        reference_forces,
+        rtol=5.0e-4,
+        atol=5.0e-6,
+    )
+    return source_atoms, target_atoms, active_distances, active_energies, active_forces
```
Comment (dallasfoster) [inline examples/neighbors/06_pair_outputs_lj.py]: So the idea of these examples are that they render well as sphinx-gallery docs and therefore the code contained within should be enlightening to a user viewing through sphinx-gallery in addition to running as a command line script. I am begrudging accepting of the allocation helper but this validation helper I find unnecessary and unhelpful as a piece of example code.

Comment (zubatyuk) [reply examples/neighbors/06_pair_outputs_lj.py]: Fixed in fc4ffea546f738a1664336fd4df38af329d91874.

File: examples/neighbors/05_jax_neighbor_list.py (hunk 3 above)
Comment (zubatyuk) [reply examples/neighbors/05_jax_neighbor_list.py]: Refined method and strategy in 4e97f455d380f56d2f6ae2ab1574db2d2b0f23af.

**Review by dallasfoster — APPROVED** [review]
